godtossing/Pos/Ssc/GodTossing/Seed.hs

```diff
@@ -6,38 +8,40 @@ module Pos.Ssc.GodTossing.Seed
-                                               CommitmentsMap (getCommitmentsMap),
-                                               OpeningsMap, SharesMap, getOpening,
-                                               secretToSharedSeed, verifyOpening)
-import           Pos.Ssc.GodTossing.Error     (SeedError (..))
-import           Pos.Ssc.GodTossing.Functions (vssThreshold)
-import           Pos.Util.Util                (getKeys)
-
+import           Control.Lens             (_Left)
+import qualified Data.HashMap.Strict      as HM
+import qualified Data.HashSet             as HS
+
+import           Pos.Binary.Class         (AsBinary, fromBinaryM)
+import           Pos.Core                 (SharedSeed, StakeholderId, addressHash, mkCoin,
+                                           sumCoins, unsafeIntegerToCoin)
+import           Pos.Crypto               (DecShare, Secret, VssPublicKey, recoverSecret,
+                                           verifySecret)
+import           Pos.Lrc.Types            (RichmenStakes)
+import           Pos.Ssc.GodTossing.Core  (Commitment (..), CommitmentsMap (..),
+                                           Opening (..), OpeningsMap, SharesMap,
+                                           SignedCommitment, getCommShares,
+                                           secretToSharedSeed, verifyOpening,
+                                           vssThreshold)
+import           Pos.Ssc.GodTossing.Error (SeedError (..))
+import           Pos.Util.Util            (getKeys)
 
 -- | Calculate SharedSeed. SharedSeed is a random bytestring that all
 -- nodes generate together and agree on.
---
--- TODO: do we need to check secrets' lengths? Probably not.
 calculateSeed
-    :: CommitmentsMap        -- ^ All participating nodes
-    -> OpeningsMap           -- ^ Openings sent by those nodes
-    -> SharesMap             -- ^ Decrypted shares
-    -> RichmenStakes          -- ^ How much stake nodes have
+    :: CommitmentsMap                         -- ^ All participating nodes
+    -> HashMap StakeholderId (AsBinary VssPublicKey) -- ^ Nodes' VSS keys
+    -> OpeningsMap                            -- ^ Openings sent by the nodes
+    -> SharesMap                              -- ^ Decrypted shares
+    -> RichmenStakes                          -- ^ How much stake nodes have
     -> Either SeedError SharedSeed
-calculateSeed commitments' openings lShares richmen = do
-    let commitments = getCommitmentsMap commitments' -- just unwrapping
-    let participants = getKeys commitments
+calculateSeed commitments' binVssKeys openings lShares richmen = do
+    let commitments :: HashMap StakeholderId SignedCommitment
+        commitments = getCommitmentsMap commitments' -- just unwrapping
+    let participants :: HashSet StakeholderId
+        participants = getKeys commitments
+    -- vssKeys :: HashMap StakeholderId VssPublicKey
+    vssKeys <- over _Left BrokenVssKey $ mapFailing fromBinaryM binVssKeys
```

Comment:
`vssKeys` is now passed because Scrape implementation doesn't guarantee ordering in some parts and thus we explicitly resolve stake holder against map, right?

Reply:
@neongreen can you explain which order we use here and how?

Reply:
SCRAPE needs to know the order of shares (i.e. you can't give it shares 4, 1 and 3, you must necessarily give it 1,3,4). It also can't deduce the order of shares because the shares don't carry any IDs or anything. Thus we need to impose some arbitrary order on them. I chose to order the shares by stakeholder's VSS key, which means that we need to know stakeholders' VSS keys in addition to a `Stakeholder->[Share]` map if we want to recover the secret.

I have put this comment into the haddock for `calculateSeed`.